File: src/main/java/top/toobee/optimization/cache/StackingMobCache.java
package top.toobee.optimization.cache;

import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.ai.memory.MemoryModuleType;
import net.minecraft.world.entity.ai.memory.NearestVisibleLivingEntities;
import net.minecraft.world.level.Level;

@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
public abstract class StackingMobCache<T extends Mob> implements AttachedCache<T> {
    private final ReentrantLock lock = new ReentrantLock();
    private final AtomicInteger referencedCounter = new AtomicInteger(0);
    private List<LivingEntity> trackers = Collections.emptyList();
    private @Nullable LivingEntity nearestTarget = null;

    protected final Level level;
    public final BlockPos pos;
    public @Nullable Boolean shouldRunMoveToTargetTask = null;
    public Optional<BlockPos> supportingBlockPos = Optional.empty();
    public Optional<LivingEntity> lookAtMobTaskEntity = Optional.empty();

    private long lastUpdateTick = 0;
    private volatile boolean recheckUpdate = false;
    private boolean hasUpdatedThisTick = false;

    public StackingMobCache(Level level, BlockPos pos) {
        this.level = level;
        this.pos = pos;
    }

    @Override public final Level getLevel() { return level; }
    @Override public final Lock getLock() { return lock; }
    @Override public final AtomicInteger getReferencedCounter() { return referencedCounter; }
    @Override public final long getLastUpdateTick() { return lastUpdateTick; }
    @Override public final boolean getHasUpdatedThisTick() { return hasUpdatedThisTick; }
    public final @Nullable LivingEntity getNearestTarget() { return nearestTarget; }
    
    @Override
    public final void setLastUpdateTick(long l) {
        this.lastUpdateTick = l;
    }

    @Override
    public final void setHasUpdatedThisTick(boolean b) {
        if (b != hasUpdatedThisTick) {
            this.hasUpdatedThisTick = b;
            this.recheckUpdate = b;
        }
    }

    public final void setTrackers(List<LivingEntity> trackers) {
        if (!recheckUpdate && lock.tryLock()) {
            try {
                this.trackers = trackers;
                this.recheckUpdate = true;
            } finally {
                lock.unlock();
            }
        }
    }

    public final void setNearestTarget(@Nullable LivingEntity e) {
        lock.lock();
        try {
            this.nearestTarget = e;
        } finally {
            lock.unlock();
        }
    }

    @Override
    public boolean failCondition(T t) {
        return !(t.blockPosition().equals(pos) && level.equals(t.level()) && t.isAlive());
    }

    @Override
    public void tick() {
        AttachedCache.super.tick();
        this.shouldRunMoveToTargetTask = null;
        this.supportingBlockPos = Optional.empty();
        this.lookAtMobTaskEntity = Optional.empty();
        this.nearestTarget = null;
        this.recheckUpdate = false;
    }

    public final void senseNearestLivingEntities(ServerLevel serverLevel, T t) {
        final var brain = t.getBrain();
        brain.setMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES, trackers);
        brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES, new NearestVisibleLivingEntities(serverLevel, t, trackers));
    }

    public abstract void newSense(ServerLevel serverLevel, T entity);
}

File: src/main/kotlin/top/toobee/optimization/cache/StackingMobCache.kt
package top.toobee.optimization.cache

import net.minecraft.entity.LivingEntity
import net.minecraft.entity.mob.MobEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.util.Optional
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock

abstract class StackingMobCache<T : MobEntity> protected constructor(
    final override val world: World,
    val pos: BlockPos,
): AttachedCache<T> {
    override val referencedCounter = AtomicInteger(0)
    override val lock = ReentrantLock()
    override var lastUpdateTick = this.world.time
    override var hasUpdatedThisTick = false
        set (b) {
            if (b != field) {
                field = b
                this.recheckUpdate.set(b)
            }
        }
    // To synchronize "hasUpdatedThisTick" of CPU cache with main memory
    private val recheckUpdate = AtomicBoolean(false)

    var shouldRunMoveToTargetTask: Boolean? = null
    var supportingBlockPos: Optional<BlockPos> = Optional.empty()
    var lookAtMobTaskEntity: Optional<LivingEntity> = Optional.empty()
    var trackers: List<LivingEntity> = emptyList()
        set (value) {
            if (!this.recheckUpdate.get() && this.lock.tryLock()) {
                try {
                    field = value
                    this.recheckUpdate.set(true)
                } finally {
                    this.lock.unlock()
                }
            }
        }
    var nearestTarget: LivingEntity? = null
        set(value) {
            lock.lock()
            try {
                field = value
            } finally {
                lock.unlock()
            }
        }

    override fun checkCondition(t: T): Boolean {
        return t.blockPos == this.pos && t.world === this.world && t.isAlive
    }

    override fun tick() {
        super.tick()
        this.shouldRunMoveToTargetTask = null
        this.supportingBlockPos = Optional.empty()
        this.lookAtMobTaskEntity = Optional.empty()
        this.nearestTarget = null
        this.recheckUpdate.set(false)
    }

    abstract fun newSense(world: ServerWorld, entity: MobEntity)

    abstract class Caches<T : MobEntity, S : StackingMobCache<T>> protected constructor(
        private val cls: Class<T>,
        protected val all: MutableMap<Pair<World?, BlockPos?>, S> = ConcurrentHashMap()
    ) {
        fun findCache(world: World?, pos: BlockPos): S? = all[world to pos]

        fun tick() = this.all.values.forEach { it.tick() }

        fun checkToCreate(world: World, pos: BlockPos, list: List<*>) {
            val count = list.filterIsInstance(cls).count { e -> e.blockPos == pos && e.world === world }
            if (count and Int.MAX_VALUE - 0b1111 != 0)
                all.computeIfAbsent(world to pos) { this.create(world, pos) }
        }

        protected abstract fun create(world: World, pos: BlockPos): S
    }
}